rfcs/20201113-tf-core.md

```diff
@@ -0,0 +1,175 @@
+
+* Is this API well tested?
+
+A well tested API, should include (1) unit tests (2) integration tests, to test all combinations and possible use cases. If it is infeasible to test some cases, it should be explicitly listed.  (3) Various hardware platforms (including CPU, GPU, TPU, etc).  Extension tests need to be included to support users to subclass things (like CompositeTensors).  Some good examples to refer to: DistStrat handles their testing strategy such that as new things are built out users automatically get tests for them as part of the standard suite invocation. It's great to know that my code won't break as APIs get extended.
+
+In the following sections, we describe what kinds of APIs will be part of Tensorflow core. This RFC describes at a high level what kinds of APIs will be a part of core without listing every possible API endpoint that we will expose. The objective is to categorize what kinds of APIs will be part of core and what kinds we will omit from Core.
+
+Note: core APIs is just a small set of existing TF APIs, they are expected to be more stable, essential, faster, better tested. All current existing TF APIs will not be changed, and will continue to be supported. 
+
+### Data structures in Core
+
+Tensor: Immutable multiple dimension denser arrays 
+Variables: Mutable shared persistent data.
+TensorArray: Dynamic-sized, write once Tensor arrays.
+Lookup tables: Hash tables 
+Element data type like tf.dtypes
+Module: named container for variables and state
+
+We’ll allow users to extend and create user defined data structures using
+ExtensionType (as known as CompositeTensor, RFC): provides a pair of protocols for defining new types that can be directly supported by TensorFlow APIs.  (This API is also known as the "CompositeTensor" API). Example ExtensionType in core includes IndexSlices.
+TypeSpec:(go/tf-type-spec) provides a standard way to define and reason about TensorFlow-specific types that are supported by TensorFlow APIs. 
+
+Note: FullType (go/tf-fulltype) definitions to be added once the design is ready.
+Note: We are designing a new full type system. Variables will inherit from core.Tensor. 
+TensorArray
+
+### The APIs on the structures mentioned above
+Creation functions for Tensor etc.
+APIs to query the attributes on Tensors include rank, shape, size.Creation
+modify, squeeze, expand Tensorflow Shape.
+get the static value from Tensor.
+Clip operation
+Concat/Split/Stack/Unstack/Pad
+Slicing APIs (scatter/gather for now but we’ll expand __getitem__ so that gather isn’t needed)
+Search/Sort/Unique/Reorder
+Transform
+Flatten and pack 
+Bitwise operations
+Conversion with other framework (TensorProto, ndarray)
+
+### Functional components of TF core
+* Control flow: tf.cond, tf.while
+* tf.function: Trace compile tensorflow programs to accelerate them 
```

Comment:
Can the quality have an impact on the evaluation about extending the core? Or do we will be forwarded on compiler improvements?